executor/adapter.go

```diff
@@ -323,6 +324,24 @@ func (a *ExecStmt) setPlanLabelForTopSQL(ctx context.Context) context.Context {
 	return topsql.AttachSQLInfo(ctx, normalizedSQL, sqlDigest, normalizedPlan, planDigest, vars.InRestrictedSQL)
 }
 
+func (a *ExecStmt) countExecForTopSQL() {
+	if !variable.TopSQLEnabled() {
+		return
+	}
+	var sqlDigest, planDigest []byte
+	vars := a.Ctx.GetSessionVars()
+	if _, d := vars.StmtCtx.SQLDigest(); d != nil {
+		sqlDigest = d.Bytes()
+	}
+	if _, d := vars.StmtCtx.GetPlanDigest(); d != nil {
+		planDigest = d.Bytes()
+	}
+	if vars.StmtStats != nil {
+		vars.StmtStats.AddExecCount(sqlDigest, planDigest, 1)
```

Comment:
When SQL duration is added later, we may need to change a lot of places like this, which means the code logic here is not flexible enough. Is it possible to be some event driven things, like `vars.StmtStats.OnStatementBeginExecution()`?
Then `SetFinished` can also be named as `OnStatementFinishExecution()`.

Reply:
@crazycs520 Is the begin position of SQLDuration the same as AddExecCount? If so, I think it would be better to add a series of event callbacks with universal names.

Reply:
Even if the needle place is not here, I think what we need to do is to add new events at that time, when existing events are not sufficient, instead of trying introduce a count everywhere possible.

Reply:
I added this interface:

```go
// StatementObserver is an abstract interface as a callback to the corresponding
// position of TiDB's SQL statement execution process. StatementStats implements
// StatementObserver and performs counting such as SQLExecCount/SQLDuration internally.
// The caller only needs to be responsible for calling different methods at the
// corresponding locations, without paying attention to implementation details.
type StatementObserver interface {
	// OnExecutionBegin should be called before statement execution.
	OnExecutionBegin(sqlDigest, planDigest []byte)

	// OnExecutionFinished should be called after the statement is executed.
	OnExecutionFinished(sqlDigest, planDigest []byte)
}
```

And call the corresponding method in different locations.